File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        List<TreeNode> res = new ArrayList<>();
        Map<String, Integer> map = new HashMap<>();
        func(root , res , map);
        return res;
    }
    String func(TreeNode root , List<TreeNode> res , Map<String , Integer> map){

        if(root ==  null){
            return "N";
        }
        String s = Integer.toString(root.val)+","+ func(root.left , res , map )+","+ func(root.right, res , map);
        map.put(s, map.getOrDefault(s, 0) + 1);
        
        if (map.get(s) == 2) {
            res.add(root);
        }
        
      return s;
    }
}

File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findDuplicateSubtrees(root *TreeNode) []*TreeNode {
    res:= []*TreeNode{}
    mapCount := make(map[string]int)
    fun(root , &res , mapCount)
    return res
}
func fun(root *TreeNode , res*[]*TreeNode , mapCount map[string]int) string{
    if root == nil {
        return "N"
    }
    s:= strconv.Itoa(root.Val)+","+fun(root.Left, res, mapCount)+","+fun(root.Right, res, mapCount)
    mapCount[s]++
    if mapCount[s] == 2{
     *res = append(*res , root)
    }
    return s
}
